public/app/plugins/datasource/graphite/specs/store.test.ts

```diff
@@ -472,4 +477,68 @@ describe('Graphite actions', async () => {
       expect(ctx.state.target.target).toEqual(expected);
     });
   });
+
+  describe('when auto-completing over a large set of tags and metrics', () => {
+    const manyMetrics: Array<{ text: string }> = [],
+      max = 20000;
+
+    beforeEach(() => {
+      for (let i = 0; i < max; i++) {
+        manyMetrics.push({ text: `metric${i}` });
+      }
+      ctx.state.datasource.metricFindQuery = jest.fn().mockReturnValue(Promise.resolve(manyMetrics));
+      ctx.state.datasource.getTagsAutoComplete = jest.fn((_tag, _prefix, { limit }) => {
+        const tags = [];
+        for (let i = 0; i < limit; i++) {
+          tags.push({ text: `tag${i}` });
+        }
+        return tags;
+      });
+      ctx.state.datasource.getTagValuesAutoComplete = jest.fn((_expression, _tag, _prefix, { limit }) => {
+        const tagValues = [];
+        for (let i = 0; i < limit; i++) {
+          tagValues.push({ text: `tagValue${i}` });
+        }
+        return tagValues;
+      });
+    });
+
+    it('uses limited metrics and tags list', async () => {
+      ctx.state.supportsTags = true;
+      const segments = await getAltSegmentsSelectables(ctx.state, 0, '');
+      expect(segments).toHaveLength(10000);
+      expect(segments[0].value!.value).toBe('*'); // * - is a fixed metric name, always added at the top
+      expect(segments[4999].value!.value).toBe('metric4998');
+      expect(segments[5000].value!.value).toBe('tag: tag0');
+      expect(segments[9999].value!.value).toBe('tag: tag4999');
+    });
+
+    it('uses correct limit for metrics and tags list when tags are not supported', async () => {
+      ctx.state.supportsTags = false;
+      const segments = await getAltSegmentsSelectables(ctx.state, 0, '');
+      expect(segments).toHaveLength(5000);
+      expect(segments[0].value!.value).toBe('*'); // * - is a fixed metric name, always added at the top
+      expect(segments[4999].value!.value).toBe('metric4998');
+    });
+
+    it('uses limited tags when editing', async () => {
+      const tags = await getTagsAsSegmentsSelectables(ctx.state, 'any');
+      expect(tags).toHaveLength(5000);
+    });
```

Comment:
given the limiting is done by the API itself for tags, I feel this is kind of a weak test as we are testing the mock implementation rather than the mock being called with the correct parameters.

shouldn't we test for `getTagsAutoComplete` getting called with the right `limit` option instead?


**EDIT:**
I see it's done in https://github.com/grafana/grafana/pull/42056/files#diff-fb7ccfa6b3057b85136a7d5e64bf0fd40c2b6627acc04a0204f45538e90a35d6R218-R229, so maybe this these is actually superfluous?